### PR TITLE
tests: filter the debug output of the spread tests

### PR DESCRIPTION
**Pull request by sergiocazzolato**
The new tool filters the debug output of the tests and saves that to the debug.log file.  This is used as part fo the github workflow to reduce the size of the logs and make is easier to read.  Then the debug.log file is uploaded as an artifact of the job.

### CODE REVIEW

File: tests/lib/tools/filter-debug-output [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    return len(parts) > 3 and \
+        match_date(parts[0]) and \
+        match_time(parts[1]) and \
+        parts[2] in OPERATIONS
+
+
+# Process spread output lines
+def process_spread_output(debug_file):
+    debug_lines = ""
+    for line in sys.stdin:
```
Comment (zyga) [inline tests/lib/tools/filter-debug-output:42]: Could you enumerate the lines and print each line we actually print with a prefix that indicates the line number in the complete log. I think it might be easier to understand when cross-referencing the summary and the full log later on.

File: tests/lib/tools/filter-debug-output [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+                # There is a Debug output right after
+                if match_debug(line):
+                    debug_lines = line
+                else:
+                    print(line.strip())
+                    debug_lines = ""
+                continue
+            else:
+                debug_lines = debug_lines + line
```
Comment (zyga) [inline tests/lib/tools/filter-debug-output:56]: This will be somewhat inefficient even though python later on tried to optimize it. I think we could have a mode where we decide to print or not each line as we see it.

The ones we see should be printed more or less like this:

> line number:line text with right-hand-side spaces stripped

File: tests/lib/tools/filter-debug-output [hunk 3]
```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+
+import argparse
```
Comment (zyga) [inline tests/lib/tools/filter-debug-output:3]: This is a great way to improve readability. Can you add a small unit test to show how it filters / not filters a short amount of example output.

I think that over time we will find weird edge cases and without tests it will be magically misbehaving.

One thing I worry about now is the internal buffering done by spread when a test is taking longer (but not failing). I don't quite know how that is handled in practice with the code below.

**Review by zyga — COMMENTED** [review]
Thanks for writing this, as described with the few comments inline I think we should make a few changes before landing this, and we definitely want to have a unit test that shows how it works in practice.

File: tests/lib/tools/filter-debug-output (hunk 1 above)
Comment (sergiocazzolato) [reply tests/lib/tools/filter-debug-output:42]: The problem I see with this is that we will see 2 times the line number in the github logs

Also to keep the same identation (regarding the start of the log line) I need to add 0s to the line number dependending on the total number of lines, for example, when we have 6500 lines in the log I'll need to add three 0 to the first 10 lines, so, once the log is fully processsed I'll need iterate the lines and autocomplete which is not performance and makes difficult to show in real time the output.

Perhaps to make easy the cross-referencing, I can add a tag in both logs.

File: tests/lib/tools/filter-debug-output (hunk 3 above)
Comment (sergiocazzolato) [reply tests/lib/tools/filter-debug-output:3]: Added a spread test. Based on the result logs in the ci I didn't see problems related to the internat buffering. I'll recheck the current results to make sure we are not loosing data.

File: tests/lib/tools/filter-debug-output (hunk 2 above)
Comment (bboozzoo) [reply tests/lib/tools/filter-debug-output:72]: to not accumulate lines you could do something like:
```
def process_debug(from_stream: io.TextIO, start_line: str, debug_count: int, debug_file) -> str:
    """Process lines from the start of the debug section until the last line,
    returns the first line right after the debug section
    """
    with open(debug_file, "a+") as outf:
        outf.write("..... some header")
        outf.write(start_line.strip())
        for line in from_stream:
            if is_debug_finished(line):
                outf.write(".... some footer")
                return line
            
            outf.write(line.strip())
        

def process_spread_output(debug_file):
    debug_num = 0
    for line in sys.stdin:
        if match_debug_start(line):
            debug_num += 1
            line = process_debug(sys.stdin, line, debug_num, debug_file)
        print(line.strip())

```

File: tests/lib/tools/filter-debug-output [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    ]
+
+def match_date(date):
+    return re.findall(r'\d{4}-\d{2}-\d{2}', date)
+
+def match_time(time):
+    return re.findall(r'\d{2}:\d{2}:\d{2}', time)
+
+# Debug line starts with 
+def match_debug(line):
```
Comment (bboozzoo) [inline tests/lib/tools/filter-debug-output:21]:
```suggestion
def match_debug_start(line):
```

File: tests/lib/tools/filter-debug-output (hunk 2 above)
Comment (sergiocazzolato) [reply tests/lib/tools/filter-debug-output]: implemented

File: tests/lib/tools/filter-debug-output [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    'Rebooting', 'Discarding', 'Allocating', 'Waiting',
+    'Allocated', 'Connecting', 'Connected', 'Sending',
+    'Error', 'Debug', 'WARNING'
+    ]
+
+def match_date(date):
+    return re.findall(r'\d{4}-\d{2}-\d{2}', date)
+
+def match_time(time):
+    return re.findall(r'\d{2}:\d{2}:\d{2}', time)
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/filter-debug-output:19]: done

File: tests/lib/tools/filter-debug-output [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+OPERATIONS = [
+    'Preparing', 'Executing', 'Restoring',
+    'Rebooting', 'Discarding', 'Allocating', 'Waiting',
+    'Allocated', 'Connecting', 'Connected', 'Sending',
+    'Error', 'Debug', 'WARNING'
+    ]
+
+def match_date(date):
+    return re.findall(r'\d{4}-\d{2}-\d{2}', date)
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/filter-debug-output:16]: done

File: tests/lib/external/snapd-testing-tools/utils/log-parser [hunk 7]
```diff
@@ -12,15 +12,649 @@ import os
...
+    if args.output:
+        reader.export_log(args.output)
+
+    reader.print_log(args.print_details, args.print_results)
+
+
+if __name__ == "__main__":
+    main()
+
+#!/usr/bin/env python3
```
Comment (zyga) [inline tests/lib/external/snapd-testing-tools/utils/log-parser:643]: Something looks wrong here.

**Review by zyga — CHANGES_REQUESTED** [review]
This PR looks somewhat wrong now (see log-parser in the middle of the file, at line 643). Can we focus on making this smaller and reviewing/landing the first branch of the batch please? I'm somewhat confused as to the relation of all the branches - this one has a lot of changes in several areas.

I would recommand that we land the library/tests for it (unit), then individual tools with spread tests for them, then the integration of the new tools into our spread workflow and lastly any miscellaneous changes like typos/corrections elsewhere.
